### ISSUE TITLE
autotools build types

### ISSUE THREAD

**Issue by ruslo** [ruslo/hunter#411]
Use `HUNTER_PACKAGE_CONFIGURATION_TYPES` to run several type of builds like Release/Debug. Add configuration specific flags like `CMAKE_CXX_FLAGS_RELEASE` or `CMAKE_CXX_FLAGS_DEBUG`.

**Comment by pretyman** [comment]
Copying the relevant information from https://github.com/ruslo/hunter/issues/407 here:

>>Yes, I did realize there is no treatment of the CMAKE_BUILD_TYPE or CMAKE_CONFIGURATION_TYPES with autotools when I was looking at the CMake properties

>To clarify scheme should use HUNTER_PACKAGE_CONFIGURATION_TYPES variable. How each build type treated is a separate question.

>>The autotools approach to deal with different build configuration types seems to use the --prefix option of ./configure to set a custom directory for each output such as release or debug directory

>CMake approach is to add postfix to the different build types, e.g. empty for Release and d for Debug. Headers should be the same. Here is example when *Config.cmake is the same for both types:

>https://github.com/forexample/package-example
>Type specific goes to *-release.cmake/*-debug.cmake. I think changing layout of binaries may lead to the problems. Need to figure out how to deal with *.pc files. I guess they are not designed so.

>>In the ideal world I think I should have one ExternalProject_Add for each CMAKE_CONFIGURATION_TYPES or CMAKE_BUILD_TYPE, is that correct?

>Yes

>>And no duplicate header files.

>Yes

**Comment by pretyman** [comment]
This what I understood with the clang toolchain in these build types:

1. Debug:
 1. Compiled with flags: `-g`
 2. Libraries filename have what is set to `CMAKE_DEBUG_POSTFIX` added as a postfix
 3. Executables filename have no change in their names, even if `CMAKE_DEBUG_POSTFIX` is set.
2. Release: 
 1. Compiled with flags: `-O3 -DNDEBUG `
  2. Libraries filename have no change in their names. A warning is issued that variable `CMAKE_DEBUG_POSTFIX` is set but not used.
  3. Executables filename has no change in their names. A warning is issued that variable `CMAKE_DEBUG_POSTFIX` is set but not used.
2. MinSizeRel: 
 1. Compiled with flags: `-Os -DNDEBUG `
  2. Libraries filename have no change in their names. A warning is issued that variable `CMAKE_DEBUG_POSTFIX` is set but not used.
  3. Executables filename has no change in their names. A warning is issued that variable `CMAKE_DEBUG_POSTFIX` is set but not used.
2. RelWithDebInfo: 
 1. Compiled with flags: `-O2 -g -DNDEBUG`
  2. Libraries filename have no change in their names. A warning is issued that variable `CMAKE_DEBUG_POSTFIX` is set but not used.
  3. Executables filename has no change in their names. A warning is issued that variable `CMAKE_DEBUG_POSTFIX` is set but not used.

It got confusing for me that Release, MinSizeRel and RelWithDebInfo are compiled with different flags, but end up with the same filename for the library. Shouldn't they be differentiated in anyway? 

My naïve point of view is that something was not properly designed here, but there could be a good reason for things to be this way.

**Comment by ruslo** [comment]
> Executables filename have no change in their names, even if CMAKE_DEBUG_POSTFIX is set

I don't know is it a bug or feature but `CMAKE_DEBUG_POSTFIX` doesn't affect executable names (only libraries). However setting `DEBUG_POSTFIX` property explicitly works:
* http://stackoverflow.com/a/20773571/2288008

> A warning is issued that variable CMAKE_DEBUG_POSTFIX is set but not used

This can be improved I guess. Need to check that generator is not of multi-configuration type and that current build type is not `Debug`. But may be it's easier to suppress it by adding dummy usage:
```cmake
if(CMAKE_DEBUG_POSTFIX)
  # suppress warning
endif()
```

> MinSizeRel ... Libraries filename have no change in their names

Because `CMAKE_MINSIZEREL_POSTFIX` used.

> Shouldn't they be differentiated in anyway?

Yes. Hunter set default posfixes in case they are empty (i.e. not set by user):
* https://github.com/ruslo/hunter/blob/a3e833ba8c9c4c0743cb82799e7c54febb52ba54/cmake/schemes/url_sha1_cmake.cmake.in#L117

> My naïve point of view is that something was not properly designed here, but there could be a good reason for things to be this way. Would you know anything about it?

In my opinion yes, there are some flaws. E.g. if posfixes are empty they should be set to some default values by CMake (not by Hunter as for now). Executable names should respect `CMAKE_*_POSTFIX` too.

> Maybe we could have them defaulted in Hunter?

See above

**Comment by pretyman** [comment]
> Yes. Hunter set default postfixes in case they are empty (i.e. not set by user):
You never disappoint Ruslan 👍 I was using your example on https://github.com/forexample/package-example straight with CMake. Thanks for clearing it up.

**Comment by pretyman** [comment]
Relevant: https://github.com/ruslo/hunter/wiki/example.hunter_configuration_types

**Comment by pretyman** [comment]
To be dealt with:

* `*.la` files from libtool
* `*.pc` files from pkg-config

**Comment by pretyman** [comment]
Ok, so these are the possible solutions and their drawbacks:

1. Compile in temporary working directory the different build types, move to the generated `*.a` or `*.so|dylib` to `PACKAGE_ROOT/lib` adding a suffix to their filenames to mimic CMake's behavior.
 - autotools projects can have hardcoded linking flags, which could make a non-`Release` build configuration link with the `Release` build, since it carries no suffix. This seems to invalidate the whole of option 1.
 - autools projects can have generate interdependent `*.dylib`, so moving them around could (?) cause RPATH problems.

2. Have each build configuration written to a prefix  `PACKAGE_ROOT/{Build Config Name}/lib` like `PACKAGE_ROOT/Release/lib`, having the Build configuration name as a parent directory of the `lib`, `include`, `bin`. Possibly even pare a parent directory for all build configurations, such as `PACKAGE_ROOT/autotools-builds/{Build Config Name}/lib`.
 - Header files could be duplicated across different build configurations. Nothing to be done with `*.pc` or `*.la` files with this point of view. Would need to add to `PKG_CONFIG_PATH` the directory `PACKAGE_ROOT/autotools-builds/{Build Config Name}/lib`.
   * Custom behavior could be added to move identical header files to `PACKAGE_ROOT/include` leaving any header files which are not identical in its own `PACKAGE_ROOT/autotools-builds/{Build Config Name}/include` directory. Pkg-config `*.pc` files would have to be modified to add to the `Cflags` line to include the `-I${prefix}/../../include` so the build configuration specific subdirectory would find the `PACKAGE_ROOT/include` directory. However I'm not sure this will scale well, as many autotools packages could be added, I found no way to modify the `*.pc` files of only the package that is being added without processing the existing `*.pc` from previously added packages.

I guess would be best that even if the build specific binaries live in separate directory trees, a PackageConfig.cmake should be created in the `PACKAGE_ROOT/lib/cmake` with all the different build types which would point to their respective `PACKAGE_ROOT/autotools-builds/{Build Config Name}/lib/libpackage.{a,so}`.

If you have any input, it would be much appreciated. Allowing duplicate header files seems to ease things quite a bit.

**Comment by ruslo** [comment]
> Have each build configuration written to a prefix PACKAGE_ROOT/{Build Config Name}/lib

Adding custom layout may break some stuff, for example:
* RPATH. By default libraries installed to `<root>/lib` and executables to `<root>/bin`. Executable's RPATH set to `../lib` so shared libraries found in `<root>/lib`. If you change it to `<root>/lib/something` then libraries will not be found.
* `Find*.cmake` or other kind of modules that expect standard layout. E.g. module may think that package installed to `<root>/include` and `<root>/lib` so path to libs is `<path-to-headers>/../lib`. You will break that too.

> autotools projects can have hardcoded linking flags, which could make a non-Release build configuration link with the Release build, since it carries no suffix

I don't know if `*.pc` support two configurations in one file. If not I guess it's not possible to do it nicely.

> so moving them around could (?) cause RPATH problems

If you keep relative paths consistent I guess everything will be OK. However renaming dependent library may break stuff.

To summarize:
Option 1 is good but there may be some general limitations that can't be "workarounded".
Option 2 is not so good but will work with workarounds.

Also I'm okay with having one build type for autotools. Main issue I think is about not adding build-type specific flags like `-O3 -DNDEBUG` or `-g`. Mixing Release autotools libraries with Debug CMake libraries **may** lead to problems but in this case you can always do `HUNTER_CONFIGURATION_TYPES=Debug` and everything should be good. It's hard to do custom schemes as fully featured as CMake. One more motivation to move to CMake :)

**Comment by pretyman** [comment]
> Find*.cmake or other kind of modules that expect standard layout. E.g. module may think that package installed to <root>/include and <root>/lib so path to libs is <path-to-headers>/../lib. You will break that too

Good call, this is no go then.

> I don't know if *.pc support two configurations in one file. If not I guess it's not possible to do it nicely.

From what I've searched, they are not, there seem to be build options for a single build only.

Ok, so after your comments I'll propose a solution 3:

 - Read `HUNTER_CONFIGURATION_TYPES` for the package, but **enforce** that it contains only **one** value, with a descriptive error message. When other build options are needed, it can be added through `hunter_config(... VERSION ... CONFIGURATION_TYPES [Release | Debug | MinSizeRel | DebWithRelInfo] )` which will generate a new config id for the directory in `HUNTER_ROOT/_Base`, not conflicting with other builds
 - Install with `--prefix=PACKAGE_ROOT`, so it installs into `PACKAGE_ROOT/lib`, `PACKAGE_ROOT/include`, etc... as expected

**Comment by ruslo** [comment]
> Read HUNTER_CONFIGURATION_TYPES for the package, but enforce that it contains only one value, with a descriptive error message

More detailed: in build scheme we should check `HUNTER_PACKAGE_CONFIGURATION_TYPES`, this variable set to `HUNTER_CONFIGURATION_TYPES` if no `hunter_config(... CONFIGURATION_TYPES ...)` specified. It mean we need to set default value to `Release` by adding `hunter_config(... CONFIGURATION_TYPES Release)` to each package with autotools scheme. Is it okay?

**Comment by pretyman** [comment]
> It means we need to set default value to Release by adding hunter_config(... CONFIGURATION_TYPES Release) to each package with autotools scheme.

I was thinking in adding the default through `hunter_configuration_types(... CONFIGURATION_TYPES Release)` in the package's `cmake/projects/Package/hunter.cmake` for each autotools project and let it be overridden by the package config file through `hunter_config(... CONFIGURATION_TYPES ...)`.

Is this what you meant?

**Comment by ruslo** [comment]
> the default through hunter_configuration_types(... CONFIGURATION_TYPES Release) in the package's cmake/projects/Package/hunter.cmake for each autotools project

Yes, you're right. This is what I mean actually.

**Comment by pretyman** [comment]
While refactoring the existing autotools based projects I found lines such as https://github.com/ruslo/hunter/blob/b3a044f9898d61bc046001bd6e80edeb99d93255/cmake/projects/Avahi/schemes/url_sha1_avahi_autotools.cmake.in#L56 :

```CMake
    CPPFLAGS
      "-I@HUNTER_INSTALL_PREFIX@/include" # find Expat
    LDFLAGS
      "-L@HUNTER_INSTALL_PREFIX@/lib" # find Expat
```

Do you think there is any problem that these were added automatically by the `hunter_autotools_project` module? I guess if this was note done before that could've been a reason, I can't remember it though...

**Comment by ruslo** [comment]
> Do you think there is any problem that these were added automatically by the hunter_autotools_project module?

No. I think it can be added automatically.

> I guess if this was note done before that could've been a reason, I can't remember it though

I remember that I was against modifying `PATH` and `PKG_CONFIG_PATH` by default, that's why we have `MODIFY_PATH` and `MODIFY_PKG_CONFIG`. I guess I'm ready to change my mind now :) Adding:
* `-I@HUNTER_INSTALL_PREFIX@/include`
* `-L@HUNTER_INSTALL_PREFIX@/lib`
* `PKG_CONFIG_PATH=@HUNTER_INSTALL_PREFIX@/lib/pkgconfig:@HUNTER_INSTALL_PREFIX@/share/pkgconfig`
* `PATH=@HUNTER_INSTALL_PREFIX@/bin:${default_path}`

seems consistent to me, feel free to remove `MODIFY_PATH`/`MODIFY_PKG_CONFIG`. We can add options for disabling of modifications of anything in case there will be some issues about it.

**Comment by pretyman** [comment]
Ok, I will treat this in a separate issue after closing this one. I want to get most of the work I can with the autotools module to add the `glfw` package which has many other Autotools packages and that would result in less updates.

**Comment by Cyberunner23** [comment]
Could this be of any use? ([url_sha1_libsodium_autogen_autotools.cmake.in](https://github.com/ruslo/hunter/blob/develop/cmake/projects/libsodium/schemes/url_sha1_libsodium_autogen_autotools.cmake.in))

**Comment by pretyman** [comment]
Can you please close this issue? 

I guess this is done
